third_party/2.7/dateutil/parser.pyi

```diff
@@ -0,0 +1,37 @@
+from typing import List, Tuple, Optional, Callable, Union, IO, Any
+from datetime import datetime
+
+__all__ = ...  # type: List[str]
+
+
+class parserinfo(object):
+    JUMP = ... # type: List[str]
+    WEEKDAYS = ... # type: List[Tuple[str, str]]
+    MONTHS = ... # type: List[Tuple[str, str]]
+    HMS = ... # type: List[Tuple[str, str, str]]
+    AMPM = ... # type: List[Tuple[str, str]]
+    UTCZONE = ... # type: List[str]
+    PERTAIN = ... # type: List[str]
+    TZOFFSET = ... # type: Dict[str, int]
+
+    def __init__(self, dayfirst: bool=..., yearfirst: bool=...) -> None: ...
+    def jump(self, name: str) -> bool: ...
+    def weekday(self, name: str) -> Union[int, None]: ...
+    def month(self, name: str) -> Union[int, None]: ...
+    def hms(self, name: str) -> Union[int, None]: ...
+    def ampm(self, name: str) -> Union[int, None]: ...
+    def pertain(self, name: str) -> bool: ...
+    def utczone(self, name: str) -> bool: ...
+    def tzoffset(self, name: str) -> Union[int, None]: ...
+    def convertyear(self, year: int) -> int: ...
+    def validate(self, res: datetime) -> bool: ...
+
+class parser(object):
+    def __init__(self, info: Optional['parserinfo']) -> None: ...
+    def parse(self, timestr: str, default: Optional[datetime],
```

Comment:
Hm... From the source it appears the input can be one of the following, in PY2:

- a stream whose read() method returns unicode
- 8-bit string (str) -- decoded to unicode and then wrapped in io.StringIO()
- unicode string (unicode) -- wrapped in io.StringIO()

It can technically also be anything else with a read() method, but in mypy we usually don't bother with duck typing in mypy. (There's an issue for an update to PEP 484 to support duck typing, but it's not currently supported: https://github.com/python/typing/issues/11.)

So my proposal for the argument type would be `Union[str, unicode, IO[str]`.

Note that there's an unfortunate distinction between `IO[str]` and `TextIO` -- so we should use the former. There's no way currently in the type system to distinguish between files open for reading or writing (or both).